File: q0543/q0543.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int currDiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfsHelper(root);
        return currDiameter;
    }

    private int dfsHelper(TreeNode root) {
        if (root == null)
            return 0;
        int l = dfsHelper(root.left);
        int r = dfsHelper(root.right);
        currDiameter = Math.max(currDiameter, l + r);
        return Math.max(l, r) + 1;
    }
}

File: q0543/q0543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0
        self.dfsHelper(root)
        return self.diameter

    def dfsHelper(self, node):
        if not node: return 0
        l = self.dfsHelper(node.left)
        r = self.dfsHelper(node.right)
        self.diameter = max(self.diameter, l + r)
        return max(l, r) + 1
